src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs

```diff
@@ -0,0 +1,178 @@
+                FindReferencesSearchOptions.Default,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return new InheritanceTargetItem(
+                inheritanceRelationshipWithOriginalMember,
+                definition);
+        }
+
+        private static async Task<InheritanceMemberItem> CreateInheritanceMemberInfoForMemberAsync(
+            Solution solution,
+            ISymbol memberSymbol,
+            int lineNumber,
+            ImmutableArray<ISymbol> implementingMembers,
+            ImmutableArray<ISymbol> implementedMembers,
+            ImmutableArray<ISymbol> overridenMembers,
+            ImmutableArray<ISymbol> overridingMembers,
+            CancellationToken cancellationToken)
+        {
+            var implementingMemberItems = await implementingMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Implementing, cancellationToken)).ConfigureAwait(false);
+            var implementedMemberItems = await implementedMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Implemented, cancellationToken)).ConfigureAwait(false);
+            var overridenMemberItems = await overridenMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Overridden, cancellationToken)).ConfigureAwait(false);
+            var overridingMemberItems = await overridingMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Overriding, cancellationToken)).ConfigureAwait(false);
+
+            return new InheritanceMemberItem(
+                lineNumber,
+                memberSymbol.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                memberSymbol.GetGlyph(),
+                implementingMemberItems.Concat(implementedMemberItems)
+                    .Concat(overridenMemberItems)
+                    .Concat(overridingMemberItems));
+        }
+
+        /// <summary>
+        /// For the <param name="memberSymbol"/>, get all the implemented symbols.
+        /// Table for the mapping between images and inheritanceRelationship
+        /// Implemented : I↓
+        /// Implementing : I↑
+        /// Overridden: O↓
+        /// Overriding: O↑
+        /// </summary>
+        private static async Task<ImmutableArray<ISymbol>> GetImplementedSymbolsAsync(
+            Solution solution,
+            ISymbol memberSymbol,
+            CancellationToken cancellationToken)
+        {
+            if (memberSymbol is INamedTypeSymbol { IsSealed: false } namedTypeSymbol)
+            {
+                var derivedTypes = await GetDerivedTypesAndImplementationsAsync(solution, namedTypeSymbol, cancellationToken).ConfigureAwait(false);
+                return derivedTypes.CastArray<ISymbol>().ToImmutableArray();
```

Comment:
you shouldn't need ToImmutableArray hee.  unless CastArray is not returning an immutable array, in which case use ImmutableArray.CastUp